cmd/gpu_fakedev/gpu_fakedev.go

```diff
@@ -302,6 +314,99 @@ func getOptions(name string) genOptions {
 	return opts
 }
 
+func makeXelinkSideCar(topology string, gpus, tiles int, connections string) {
+	if topology != fullyConnected {
+		log.Printf("XELINK: generate xelink sidecar label file, using (GPUs: %d, Tiles: %d)", gpus, tiles)
+	} else {
+		log.Printf("XELINK: generate xelink sidecar label file, using (GPUs: %d, Tiles: %d, Topology: %s)", gpus, tiles, topology)
+	}
+
+	if topology == fullyConnected {
+		saveSideCarFile(buildConnectionList(gpus, tiles))
+	} else {
+		saveSideCarFile(connections)
+	}
+}
+
+func buildConnectionList(gpus, tiles int) string {
+	mm := 0
+
+	var nodes = make([]string, 0)
+
+	for mm < gpus {
+		nn := 0
+		for nn < tiles {
+			nodes = append(nodes, fmt.Sprintf("%d.%d", mm, nn))
+			nn++
+		}
+		mm++
+	}
+
+	var links = make(map[string]struct{}, 0)
+
+	var smap = make([]string, 0)
+
+	for _, from := range nodes {
+		for _, to := range nodes {
+			// no self links, TODO ignore in-gpu xelinks
+			if to == from {
+				continue
+			}
+
+			link := fmt.Sprintf("%s-%s", to, from)
+
+			reverselink := fmt.Sprintf("%s-%s", from, to)
+			if _, exists := links[reverselink]; !exists {
+				links[link] = struct{}{}
+
+				smap = append(smap, link)
+			}
+		}
+	}
+
+	return fmt.Sprintf("%s\n", strings.Join(smap, "_"))
+}
+
+func saveSideCarFile(connections string) {
+	f, err := os.Create("xpum-sidecar-labels.txt")
+	if err != nil {
+		panic(err)
+	}
+	defer f.Close()
+
+	if len(connections) <= maxK8sLabelSize {
+		line := fmt.Sprintf("xpumanager.intel.com/xe-links=%s", connections)
+		fmt.Println(line)
+
+		if _, err := f.WriteString(line + "\n"); err != nil {
+			panic(err)
+		}
+
+		return
+	}
+
+	// Write first line without Z prefix
+	line := fmt.Sprintf("xpumanager.intel.com/xe-links=%s", connections[:maxK8sLabelSize])
+	fmt.Println(line)
+
+	if _, err := f.WriteString(line + "\n"); err != nil {
+		panic(err)
+	}
```

Comment:
Sorry for a late comment. I was re-reading this and I think these two could be combined. It's essentially same thing twice. Even the part below could be combined, but that would probably get too confusing.